dist/ExtUtils-ParseXS/lib/perlxs.pod

```diff
@@ -1,2065 +1,4706 @@
+the C<printf> above). The ellipsis is added to ignore the extra two
+arguments passed to an overload method.
+
+The original C<mynum_add()> method which, via aliasing, handled all four
+of the arithmetic operations, is now split into four separate XSUBs, since
+C<ALIAS> and C<OVERLOAD> doesn't mix.
+
+The main change to each arithmetic XSUB part from adding the C<OVERLOAD>
+keyword, is that there is an extra C<swap> parameter. There's no real need
+to use it for addition and multiplication, but it is important for the
+non-commutative subtraction and division operations.
+
+That example uses the C<T_PTROBJ> typemap to process the second argument,
+which in the most general usage may not be an object. For example the
+second and third of these lines will croak with an C<Expected foo to be of
+type My::Num, got scalar> error:
+
+    $i13 + My::Num->new(7);
+    $i13 + 7;
+    $i13 + "7";
+
+If it is necessary to handle this, then you may need to create your own
+typemap: for example, something similar to C<T_PTROBJ>, but with an INPUT
+template along the lines of:
+
+    T_MYNUM
+        SV *sv = $arg;
+        SvGETMAGIC(sv);
+        if (!SvROK(sv)) {
+            sv = sv_newmortal();
+            sv_setref_pv(sv, "$ntype", mynum_new(SvIV($arg));
+        }
+        ....
 
-  void
-  day_month(day, unix_time, month)
-      int &day = NO_INIT
-      int  unix_time
-      int &month = NO_INIT
-    OUTPUT:
-      day
-      month
+Finally, although not directly related to XS, the following could be added
+to F<Num.pm> to allow integer literals to be used directly:
 
-However, the generated Perl function is called in very C-ish style:
+    sub import {
+        overload::constant integer =>
+            sub {
+                my $str = shift;
+                return My::Num->new($str);
+            };
+    }
 
-  my ($day, $month);
-  day_month($day, time, $month);
+which then allows these lines:
 
-=head2 The C<length(NAME)> Keyword
+    my $i2  = My::Num->new(2);
+    my $i7  = My::Num->new(7);
+    my $i13 = My::Num->new(13);
 
-If one of the input arguments to the C function is the length of a string
-argument C<NAME>, one can substitute the name of the length-argument by
-C<length(NAME)> in the XSUB declaration.  This argument must be omitted when
-the generated Perl function is called.  E.g.,
+to be rewritten more cleanly as:
 
-  void
-  dump_chars(char *s, short l)
-  {
-    short n = 0;
-    while (n < l) {
-        printf("s[%d] = \"\\%#03o\"\n", n, (int)s[n]);
-        n++;
-    }
-  }
+    my $i2  = 2;
+    my $i7  = 7;
+    my $i13 = 13;
 
-  MODULE = x		PACKAGE = x
+=head3 The ATTRS: Keyword
 
-  void dump_chars(char *s, short length(s))
+    MODULE = Foo::Bar PACKAGE = Foo::Bar
 
-should be called as C<dump_chars($string)>.
+    SV*
+    debug()
+      ATTRS: lvalue
+      PPCODE:
+        # return $Foo::Bar::DEBUG, creating it if not already present:
+        PUSHs(GvSV(gv_fetchpvs("Foo::Bar::DEBUG", GV_ADD, SVt_IV)));
```

Comment:
Woah, why doesn't this need to extend the stack first? If it doesn't an explanatory comment might help others with the same question

Reply:
There's always room for one element on the stack (the spot the CV was popped from in pp_entersub).

I experimented with requiring the extend for 1 argument returns in the HWM check: it broke a lot of things.

Reply:
Ah, I recall coming across that before! Right... I think this could definitely use a comment then